File: leetcode/python/1392.longest-happy-prefix.py
class Solution:
    def longestPrefix(self, s: str) -> str:
        prefixLen = self.calcPrefixLen(s)
        # print(prefixLen)
        return s[:prefixLen[-1]]
    

    ## calculating KMP prefix arr
    def calcPrefixLen(self, pattern: str) -> List[int]:
        patternLen = len(pattern)
        arr = [0] * (patternLen + 1)
        arr[0] = -1

        prefixLen = 0
        i = 1

        while i < patternLen:
            if pattern[prefixLen] == pattern[i]:
                prefixLen += 1
                i += 1
                arr[i] = prefixLen
            elif prefixLen > 0:
                prefixLen = arr[prefixLen]
            else:
                i += 1
                arr[i] = 0
        
        return arr

File: leetcode/cpp/1392.longest-happy-prefix.cpp
class Solution {
public:
    string longestPrefix(string s) {
        auto prefixLen = calcPrefixLen(s);
        return s.substr(0, prefixLen.back());
    }
private:
    // calculating KMP prefix arr
    vector<int> calcPrefixLen(string& pattern) {
        int patternLen = pattern.length();

        vector<int> arr(patternLen+1, 0);
        arr[0] = -1;

        int prefixLen = 0;
        int i = 1;

        while (i < patternLen) {
            if (pattern[i] == pattern[prefixLen]) {
                prefixLen++;
                i++;
                arr[i] = prefixLen;
            } else if (prefixLen > 0) {
                prefixLen = arr[prefixLen];
            } else {
                i++;
                arr[i] = 0;
            }
        }
        return arr;
    }
};

